README
NAME
    Schema::Data - Abstract class for versioned DB schema datasets.

SYNOPSIS
     package Schema::Data::Foo;
     use base qw(Schema::Data);

     sub _versions_file {
             return 'versions.txt';
     }

     package main;

     my $obj = Schema::Data::Foo->new;
     my @versions = $obj->list_versions;
     my $schema_data = $obj->schema_data;
     my $version = $obj->version;

DESCRIPTION
    This is abstract class for versioned schema dataset modules.

METHODS
  "new"
     my $obj = Schema::Data->new(%params);

    Constructor.

    Returns instance of object.

    *       "version"

            Selected schema data version.

            Default value is last version.

  "list_versions"
     my @versions = $obj->list_versions;

    Get sorted list of versions.

    Returns array of versions.

  "schema_data"
     my $schema_data = $obj->schema_data;

    Get schema data module name.

    Returns string.

  "version"
     my $version = $obj->version;

    Get version of schema data, which is actial set.

    Returns string.

ERRORS
     new():
             Cannot load Schema data module.
                     Module name: %s
                     Error: %s
             Schema data version has bad format.
                     Schema data version: %s
             From Class::Utils::set_params():
                     Unknown parameter '%s'.

             (only in this abstract class)
             We need to implement distribution file with Schema data versions.

EXAMPLE
     use strict;
     use warnings;

     use File::Path qw(make_path);
     use File::Spec::Functions qw(catfile);
     use File::Temp qw(tempdir tempfile);
     use IO::Barf qw(barf);

     # Temp directory for generated module
     my $temp_dir = tempdir(CLEANUP => 1);

     # File with versions.
     my (undef, $versions_file) = tempfile();

     make_path(catfile($temp_dir, 'Schema', 'Data', 'Foo'));

     my $package_schema_foo = catfile($temp_dir, 'Schema', 'Data', 'Foo.pm');
     barf($package_schema_foo, <<"END");
     package Schema::Data::Foo;

     use base qw(Schema::Data);

     use IO::Barf qw(barf);

     sub _versions_file {
             barf('$versions_file', "0.2.0\\n0.1.0\\n0.1.1");

             return '$versions_file';
     }

     1;
     END
 
     my $package_schema_foo_0_1_0 = catfile($temp_dir, 'Schema', 'Data', 'Foo', '0_1_0.pm');
     barf($package_schema_foo_0_1_0, <<'END');
     package Schema::Data::Foo::0_1_0;

     1;
     END

     my $package_schema_foo_0_1_1 = catfile($temp_dir, 'Schema', 'Data', 'Foo', '0_1_1.pm');
     barf($package_schema_foo_0_1_1, <<'END');
     package Schema::Data::Foo::0_1_1;

     1;
     END

     my $package_schema_foo_0_2_0 = catfile($temp_dir, 'Schema', 'Data', 'Foo', '0_2_0.pm');
     barf($package_schema_foo_0_2_0, <<'END');
     package Schema::Data::Foo::0_2_0;

     1;
     END

     unshift @INC, $temp_dir;

     require Schema::Data::Foo;

     my $obj = Schema::Data::Foo->new;

     my @versions = $obj->list_versions;

     print join "\n", @versions;

     unlink $versions_file;

     # Output:
     # 0.1.0
     # 0.1.1
     # 0.2.0

DEPENDENCIES
    Class::Utils, English, Error::Pure, Perl6::Slurp.

REPOSITORY
    <https://github.com/michal-josef-spacek/Schema-Data>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

LICENSE AND COPYRIGHT
    © 2022-2023 Michal Josef Špaček

    BSD 2-Clause License

VERSION
    0.07



Changes
0.07

0.06 2023-03-01T00:51:30+01:00
 - Add LICENSE file.
 - Add README file.
 - Update copyright years.

0.05 2022-10-20T11:19:03+02:00
 - Add schema() method to Schema::Data::Data abstract class.
 - Add Schema::Data::Plugin abstract class.

0.04 2022-10-04T23:13:21+02:00
 - Add 'db_debug' parameter to Schema::Data::Data.
 - Fix Schema::Data::Data user and password setting.

0.03 2022-09-09T17:06:05+02:00
 - Add variables for insert() in Schema::Data::Data.

0.02 2022-09-09T01:21:56+02:00
 - Fix minimal dependency for List::Util.

0.01 2022-09-08T01:50:32+02:00
 - First version.


Data/Data.pm
package Schema::Data::Data;

use strict;
use warnings;

use Class::Utils qw(set_params);
use English;
use Error::Pure qw(err);
use List::Util qw(none);

our $VERSION = 0.07;

sub new {
	my ($class, @params) = @_;

	# Create object.
	my $self = bless {}, $class;

	# db debug.
	$self->{'db_debug'} = 0;

	# db user password.
	$self->{'db_password'} = undef;

	# db user name.
	$self->{'db_user'} = undef;

	# db options.
	$self->{'db_options'} = {};

	# DSN.
	$self->{'dsn'} = undef;

	# Schema module.
	$self->{'schema_module'} = undef;

	# Process parameters.
	set_params($self, @params);

	if (! defined $self->{'schema_module'}) {
		err "Parameter 'schema_module' is required.";
	}

	# Load schema.
	eval 'require '.$self->{'schema_module'};
	if ($EVAL_ERROR) {
		err 'Cannot load Schema module.',
			'Module name', $self->{'schema_module'},
			'Error', $EVAL_ERROR,
		;
	}

	$self->{'_schema'} = eval {
		$self->{'schema_module'}->connect(
			$self->{'dsn'},
			$self->{'db_user'},
			$self->{'db_password'},
			$self->{'db_options'},
		);
	};
	if ($EVAL_ERROR) {
		err 'Cannot connect to Schema database.',
			'Error', $EVAL_ERROR,
		;
	}
	if (! $self->{'_schema'}->isa('DBIx::Class::Schema')) {
		err "Instance of schema must be a 'DBIx::Class::Schema' object.",
			'Reference', $self->{'_schema'}->isa,
		;
	}

	if ($self->{'db_debug'}) {
		$self->{'_schema'}->storage->debug(1);
	}

	return $self;
}

sub data {
	my ($self, $variables_hr) = @_;

	err 'Package __PACKAGE__ is abstract class. data() method must be '.
		'defined in inherited class.';
}

sub insert {
	my ($self, $variables_hr) = @_;

	# Check required variables.
	foreach my $required_variable ($self->required_variables) {
		if (! exists $variables_hr->{$required_variable}) {
			err "Variable '$required_variable' is required.";
		}
	}

	my $data_hr = $self->data($variables_hr);
	my %sources = map { ($self->{'_schema'}->source($_)->name => $_ ) }
		$self->{'_schema'}->sources;

	# Explicit order.
	my @order;
	if ($data_hr->{'_order'}) {
		@order = @{$data_hr->{'_order'}};
		foreach my $db_name (keys %sources) {
			if (none { $_ eq $db_name } @order) {
				push @order, $db_name;
			}
		}
	# Random DB names as order.
	} else {
		@order = keys %sources;
	}

	# Insert.
	foreach my $db_name (@order) {

		# Skip db table without data.
		if (! exists $data_hr->{$db_name}) {
			next;
		}

		foreach my $data_hr (@{$data_hr->{$db_name}}) {
			$self->{'_schema'}->resultset($sources{$db_name})->create($data_hr);
		}
	}

	return;
}

sub required_variables {
	my $self = shift;

	err 'Package __PACKAGE__ is abstract class. required_variables() method must be '.
		'defined in inherited class.';
}

sub schema {
	my $self = shift;

	return $self->{'_schema'};
}

1;

__END__


Data/Plugin.pm
package Schema::Data::Plugin;

use strict;
use warnings;

use Class::Utils qw(set_params);
use Error::Pure qw(err);
use Scalar::Util qw(blessed);

our $VERSION = 0.07;

sub new {
	my ($class, @params) = @_;

	# Create object.
	my $self = bless {}, $class;

	# Schema.
	$self->{'schema'} = undef;

	# Verbose callback.
	$self->{'verbose_cb'} = undef;

	# Process parameters.
	set_params($self, @params);

	if (! defined $self->{'schema'}) {
		err "Parameter 'schema' is required.";
	}
	if (! blessed($self->{'schema'}) || ! $self->{'schema'}->isa('DBIx::Class::Schema')) {
		err "Parameter 'schema' must be a instance of 'DBIx::Class::Schema'.";
	}

	if (defined $self->{'verbose_cb'} && ref $self->{'verbose_cb'} ne 'CODE') {
		err "Parameter 'verbose_cb' must be reference to code.";
	}

	return $self;
}

sub load {
	my $self = shift;

	err 'Package __PACKAGE__ is abstract class. load() method must be '.
		'defined in inherited class.';
}

sub supported_versions {
	my $self = shift;

	err 'Package __PACKAGE__ is abstract class. supported_versions() method must be '.
		'defined in inherited class.';
}

1;

__END__


Makefile.PL
use lib '.';
use strict;
use warnings;

use inc::Module::Install;

# Definition.
abstract 'Abstract class for classes for database data.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.05;
author_requires 'Test::More' => 0;
author_requires 'Test::NoWarnings' => 0;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
license 'bsd';
name 'Schema-Data';
perl_version 5.006;
readme_from 'Data.pm';
recursive_author_tests('xt');
requires 'Class::Utils' => 0;
requires 'English' => 0;
requires 'Error::Pure' => 0;
requires 'List::Util' => 1.33;
requires 'Perl6::Slurp' => 0;
requires 'Scalar::Util' => 0;
resources 'bugtracker' => 'https://github.com/michal-josef-spacek/Schema-Data/issues',
resources 'homepage' => 'https://github.com/michal-josef-spacek/Schema-Data';
resources 'repository' => 'git://github.com/michal-josef-spacek/Schema-Data';
test_requires 'English' => 0;
test_requires 'File::Object' => 0;
test_requires 'File::Spec::Functions' => 0;
test_requires 'File::Temp' => 0;
test_requires 'IO::Barf' => 0;
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
tests_recursive;
version '0.07';

# Run.
WriteAll();
